src/hotspot/share/opto/vectornode.hpp

```diff
@@ -216,20 +232,44 @@ class ReductionNode : public Node {
   virtual uint size_of() const { return sizeof(*this); }
 };
 
+//---------------------------UnorderedReductionNode-------------------------------------
+// Order of reduction does not matter. Example int add. Not true for float add.
+class UnorderedReductionNode : public ReductionNode {
+public:
+  UnorderedReductionNode(Node * ctrl, Node* in1, Node* in2) : ReductionNode(ctrl, in1, in2) {
+    init_class_id(Class_UnorderedReduction);
+  }
+
+  virtual VectorNode* make_normal_vector_op(Node* in1, Node* in2, const TypeVect* vt) = 0;
+  virtual bool make_normal_vector_op_implemented(const TypeVect* vt) = 0;
```

Comment:
How about introducing `virtual int vect_Opcode()` (`norm_vect_Opcode()`) or something which returns normal vector opcode (`Op_AddVI` for `AddReductionVINode` for example). Then you don't need these 2 functions to be virtual:
```
  virtual int vect_Opcode() const = 0;
  VectorNode* make_normal_vector_op(Node* in1, Node* in2, const TypeVect* vt) {
    return new VectorNode::make(vect_Opcode(), in1, in2, vt);
  }
  bool make_normal_vector_op_implemented(const TypeVect* vt) {
    return Matcher::match_rule_supported_vector(vect_Opcode(), vt->length(), vt->element_basic_type());
  }
```

If we need that in more cases then in your changes may be have even more general (in `VectorNode` class) `scalar_Opcode()` and use `VectorNode::opcode(sclar_Opcode(), vt->element_basic_type())` to get normal vector opcode. This may need more changes and testing - a separate RFE.

Reply:
I am also not sure about need `_op` in these functions names.